Doc on expressions : move and complete expressions [ANT-4117] (#3217)
AntaresSimulatorTeam/Antares_Simulator: docs/user-guide/modeler/09-expressions.md

They allow a large flexibility in defining objects of optimization.

Expressions can only be found in [models](../02-inputs.md#Models) (defined in libraries yaml files).

More specifically, they can be found in the following contexts inside a model : 
- **constraints**
- **binding-constraints**
- **objective-contributions**
- **port-field-definitions**
- in fields **lower-bound** and **upper-bound** of **variables**.
- **extra-output**

Many elements can be found in an expression. In the following sections, we give a list of these elements, but have in mind that all these elements cannot be used in any kind of expression.
It depends on the context of the expression inside the model.
For instance, if the expression is used to define the **lower-bound** of a **variable**, it cannot contain a reference to any **variable**.
For more precision see [Cases where elements can be used in expressions]

Here is the list of elements possibly composing an expression :

## Arithmetic operators in expressions

The following operators are allowed between two elements :

- **+**: addition
- **-**: subtraction

> - __*__: multiplication (muliplying 2 variables is always forbidden)

Reviewer: It's actually allowed in extra-outputs, please update
```suggestion
- __*__: multiplication
```

- **/**: division (dividing 2 variables is always forbidden)
- **=**: equality, only allowed for constraint definitions
- **<=**: less or equal to, only allowed for constraint definitions

> - **>=**: greater or equal to, only allowed for constraint definitions

Reviewer 2: unless extra-output

```

### Power operator

This unary operator **^** is used within an expression whose context is an extra output.
It can only apply to a reference to an existing variable.

_Example :_ 

```yaml
models:
  - id: myModel
    variables:
    - id: myVar
    parameters:
    - id: myParam
    extra-outputs:
    - id: myOutput
      expression: myVar^(2 + myParam)
```

### Max operator

This binary operator **max(u, v)** can only be used inside an expression whose context is either **constraints** or **extra-output**.
- in the context of a **constraint** (or a **binding constraint**), the 2 operands must be parameters or literals.
- in the context of an **extra output**, there are no restriction on operands.

_Example :_

```yaml
models:
  - id: myModel
    variables:
    - id: x
    parameters:
    - id: a
    - id: b
    constraints:
    - id: myConstraint
      expression: x < max(a,b)
    extra-outputs:
    - id: myOutput
      expression: max(x, a*b)
```

## Cases where elements can be used in expressions

### Operators

**Caution** : multiplying or dividing two references to variables is always forbidden.
Variables can only be multiplied/divided by literals/parameters. 

> These restrictions are implicit in the following tables.

Reviewer: ```suggestion
These restrictions are explicited in the following tables.
```

Author: Done

---

|Context of expression        | [+-]  | [*/] | [<>=] |  Time | sum_connections |
|-----------------------------|-------|------|-------|-------|-----------------|
|constraints                  |  x    |   x  |  x    |   x   |  x              |

> |binding-constraints          |  x    |   x  |  x    |   x   |  x              |

Reviewer: Constraints/binding-constraints actually **require** one of <=>
```suggestion
|constraints                  |  x    |   x  |  o    |   x   |  x              |
|binding-constraints          |  x    |   x  |  o    |   x   |  x              |
```
Also, `variable/2` is perfectly valid in all contexts, but `2/variable` is only valid in extra-outputs.

|objective-contributions      |  x    |   o  |  o    |   o   |  o              |
|port-field-definitions       |  x    |   x  |  o    |   o   |  o              |
|variable bounds              |  x    |   x  |  o    |   o   |  o              |
|extra-output                 |  x    |   x  |  o    |   o   |  o              |

---

|Context of expression        | Scenario | Dual | Power | Max | sum |
|-----------------------------|----------|------|-------|-----|-----|
|constraints                  |    ?     |   o  |   o   |  x  |  x  |
|binding-constraints          |    ?     |   o  |   o   |  x  |  x  |
|objective-contributions      |    ?     |   o  |   o   |  o  |  x  |
|port-field-definitions       |    ?     |   o  |   o   |  o  |  x  |
|variable bounds              |    ?     |   o  |   o   |  o  |  x  |

> |extra-output                 |    ?     |   x  |   x   |  x  |  x  |

Reviewer: What is "Scenario" ? I'd suggest removing this column for now.

```suggestion
|constraints                  |    ?     |   x  |   NV   |  NV  |  x  |
|binding-constraints          |    ?     |   x  |   NV   |  NV  |  x  |
|objective-contributions      |    ?     |   x  |   NV   |  NV  |  x  |
|port-field-definitions       |    ?     |   x  |   NV   |  NV  |  x  |
|variable bounds              |    ?     |   x  |   NV   |  NV  |  x  |
|extra-output                 |    ?     |   o  |   o   |  o  |  o  |

NV : can be applied on "non-variable" elements __ie__ all elements except variables
```

### References to elements defined elsewhere

|Context of expression        | variable  | parameter| Port |
|-----------------------------|-----------|----------|------|
|constraints                  |  x        |   x      |  x   |

> |binding-constraints          |  x        |   x      |  x   |

Reviewer: The whole point of binding-constraints is to plug port fields
```suggestion
|binding-constraints          |  x        |   x      |  o   |
```

|objective-contributions      |  x        |   x      |  o   |
|port-field-definitions       |  x        |   x      |  o   |

> |variable bounds              |  o        |   x      |  o   |

Reviewer: ```suggestion
|objective-contributions      |  x        |   x      |  x   |
|port-field-definitions       |  x        |   x      |  x   |
|variable bounds              |  o        |   x      |  x   |
```
I think port fields can only appear in binding-constraints and extra-outputs

- in the context of a **constraint** (or a **binding constraint**), the 2 operands must be parameters or literals.
- in the context of an **extra output**, there are no restriction on operands.

_Example :_

```yaml
models:
  - id: myModel
    variables:
    - id: x
    parameters:
    - id: a
    - id: b
    constraints:
    - id: myConstraint
      expression: x < max(a,b)
    extra-outputs:
    - id: myOutput
      expression: max(x, a*b)
```

## Cases where elements can be used in expressions

### Operators

**Caution** : as already said, non-linear multiplications or divisions are forbidden in expressions in general, except within an **extra-output** expressions.
Apart from that case, variables can only be multiplied/divided by literals/parameters. 

In following tables : 
- **L** means : only linear multiplication or division is allowed
- **NL** means non linear multiplication or division is allowed

---

|Context of expression        | [+-]  | [*/] | [<>=] |  Time | sum_connections |
|-----------------------------|-------|------|-------|-------|-----------------|
|constraints                  |  yes  |   L  |  yes  |  yes  |  yes              |

> |binding-constraints          |  yes  |   L  |  yes  |  yes  |  yes              |

Reviewer: ```suggestion
|constraints                  |  yes  |   L  |  yes  |  yes  |  no              |
|binding-constraints          |  yes  |   L  |  yes  |  yes  |  yes              |
```

Author: Done

~~~

### Dual operators

In some cases, we need to access dual results of variables / constraints of the linear problem.
Depending on the case, the dual unary operator is : 
- dual result of a **variable** whose id is **my_var** is accessed by **-reduced_cost(myVar)**
- dual result of a **constraint** whose if is **my_constraint** is accessed by **dual(myConstraint)**

Note that dual results can only used within an expression whose context is an extra output. 

_Exemple_ : 

```yaml
models:
  - id: myModel
    variables:
    - id: myVar
      upper_bound: 1
    constraints:
    - id: myConstraint
      expression: x <= 1
    extra-outputs:
    - id: marginal_price_variable
      expression: -reduced_cost(myVar)
    - id: marginal_price_constraint
      expression: dual(myConstraint)
```

### Power operator

This binary operator **^** is used within an expression whose context is an extra output.
It can only apply to a reference to an existing variable.

_Example :_ 

```yaml
models:
  - id: myModel
    variables:
    - id: myVar
    parameters:
    - id: myParam
    extra-outputs:
    - id: myOutput
      expression: myVar^(2 + myParam)
```

### Max (or Min) operator

> This binary operator **max(u, v)** can only be used inside an expression whose context is either **constraints** or **extra-output**.

Reviewer 2: ```suggestion
This n-ary operator **max(u, v, etc...)** can only be used inside an expression whose context is either **constraints** or **extra-output**.
```

### Operators

**Caution** : multiplying or dividing two references to variables is always forbidden.

> Variables can only be multiplied/divided by literals/parameters. 

Reviewer 2: unless extra-output
